### PR TITLE
implement new interrupt calling conventions in the llvm backend

### LINKED ISSUE #20671: Unable to use `callconv(.Interrupt)` for riscv targets
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

Currently compiling this function for riscv targets will yield the following error:
```zig
export fn _irq() callconv(.Interrupt) void {
    asm volatile ("nop");
}
```
```
error: callconv 'Interrupt' is only available on x86, x86_64, AVR, and MSP430, not riscv32
```

irq routines in riscv requires a different `mret` instruction at the end and this is achieved in C as in the following:
```c
void _irq() __attribute__((interrupt)) {
    asm ("nop");
}
```
for which LLVM generates the following code:
```
000111e0 <_irq>:
   111e0:       1141                    addi    sp,sp,-16
   111e2:       c606                    sw      ra,12(sp)
   111e4:       c422                    sw      s0,8(sp)
   111e6:       0800                    addi    s0,sp,16
   111e8:       0001                    nop
   111ea:       40b2                    lw      ra,12(sp)
   111ec:       4422                    lw      s0,8(sp)
   111ee:       0141                    addi    sp,sp,16
   111f0:       30200073                mret
```

I dont know enough about zig/llvm internals to know if the zig `callconv(.Interrupt)` has anything to do with the way C's attributes are handled wrt llvm but it'll be useful to be able to implement riscv irq routines in pure zig.

### Expected Behavior

Able to use `callconv(.Interrupt)` on riscv targets.

### LINKED ISSUE #21805: Remove default field value for `std.builtin.CallingConvention.RiscvInterruptOptions.mode`
Extracted from #21697. The justification is in that PR.

cc @Rexicon226

### PR DESCRIPTION
**Pull request by Rexicon226**
closes #20671
closes #21805 

anyone know of other issues this closes?

adds support for:
- `riscv64_interrupt`
- `riscv32_interrupt`
- `mips_interrupt`
- `mips64_interrupt`
- `arm_interrupt`
- `csky_interrupt`

### CODE REVIEW

**Comment by alexrp** [comment]
#16635 can maybe also be closed? Though not as a result of this PR specifically.

File: src/codegen/llvm.zig [hunk 1]
```diff
@@ -3076,6 +3076,54 @@ pub const Object = struct {
...
+                    try attributes.addFnAttr(.{ .string = .{
+                        .kind = try o.builder.string(switch (tag) {
+                            .avr_interrupt => "interrupt",
+                            .avr_signal => "signal",
+                            else => unreachable,
+                        }),
+                        .value = .empty,
+                    } }, &o.builder);
+                },
+                .csky_interrupt => {
```
Comment (alexrp) [inline src/codegen/llvm.zig:3119]: I guess you could merge this with the prong above? Doesn't really matter though.

**Review by alexrp — APPROVED** [review]

File: src/codegen/llvm.zig [hunk 2]
```diff
@@ -11609,9 +11657,13 @@ pub fn toLlvmCallConv(cc: std.builtin.CallingConvention, target: std.Target) ?Ca
...
             void => .{ null, 0 },
-            std.builtin.CallingConvention.CommonOptions => .{ pl.incoming_stack_alignment, 0 },
+            std.builtin.CallingConvention.ArmInterruptOptions,
+            std.builtin.CallingConvention.RiscvInterruptOptions,
+            std.builtin.CallingConvention.MipsInterruptOptions,
+            std.builtin.CallingConvention.CommonOptions,
+            => .{ pl.incoming_stack_alignment, 0 },
             std.builtin.CallingConvention.X86RegparmOptions => .{ pl.incoming_stack_alignment, pl.register_params },
-            else => unreachable,
+            else => @compileError("TODO: toLlvmCallConv" ++ @tagName(pl)),
```
Comment (mlugg) [inline src/codegen/llvm.zig:11666]: nit: don't try and be too fancy, `comptime unreachable` is fine :^)

File: src/codegen/llvm.zig [hunk 3]
```diff
@@ -3076,6 +3076,54 @@ pub const Object = struct {
...
+                inline .avr_interrupt, .avr_signal => |_, tag| {
+                    try attributes.addFnAttr(.{ .string = .{
+                        .kind = try o.builder.string(switch (tag) {
+                            .avr_interrupt => "interrupt",
+                            .avr_signal => "signal",
+                            else => unreachable,
+                        }),
+                        .value = .empty,
+                    } }, &o.builder);
+                },
```
Comment (mlugg) [inline src/codegen/llvm.zig:3118]: Don't use `inline` unnecessarily:
```suggestion
                .avr_interrupt, .avr_signal => {
                    try attributes.addFnAttr(.{ .string = .{
                        .kind = try o.builder.string(switch (fn_info.cc) {
                            .avr_interrupt => "interrupt",
                            .avr_signal => "signal",
                            else => unreachable,
                        }),
                        .value = .empty,
                    } }, &o.builder);
                },
```

File: src/codegen/llvm.zig (hunk 2 above)
Comment (Rexicon226) [reply src/codegen/llvm.zig:11666]: yeah fair enough, I just needed to see which callconv was the one I hadn't handled yet and I'm lazy. would you like me to use `comptime unreachable` here instead?

Comment (mlugg) [reply src/codegen/llvm.zig:11666]: Feel free to change it, but it's not a problem, just thought I'd comment

File: src/codegen/llvm.zig (hunk 3 above)
Comment (Rexicon226) [reply src/codegen/llvm.zig]: Fixed, and combined with Alex's suggestion of putting `csky_interrupt` into there as well.

**Pull request merged**